samza-test/src/test/java/org/apache/samza/test/processor/TestZkLocalApplicationRunner.java

```diff
@@ -0,0 +1,397 @@
+      try {
+        LOGGER.info("Publish kafka event with index : {} for stream processor: {}.", eventIndex, streamProcessorId);
+        producer.send(new ProducerRecord(topic, new TestKafkaEvent(streamProcessorId, String.valueOf(eventIndex)).toString().getBytes()));
+      } catch (Exception  e) {
+        LOGGER.error("Publishing to kafka topic: {} resulted in exception: {}.", new Object[]{topic, e});
+        throw new SamzaException(e);
+      }
+    }
+  }
+
+  private ApplicationConfig buildStreamApplicationConfig(String systemName, String inputTopic,
+                                                         String processorId, String appName, String appId) {
+    Map<String, String> samzaContainerConfig = ImmutableMap.<String, String>builder()
+        .put(TaskConfig.INPUT_STREAMS(), inputTopic)
+        .put(JobConfig.JOB_DEFAULT_SYSTEM(), systemName)
+        .put(TaskConfig.IGNORED_EXCEPTIONS(), "*")
+        .put(ZkConfig.ZK_CONNECT, zkConnect())
+        .put(JobConfig.SSP_GROUPER_FACTORY(), TEST_SSP_GROUPER_FACTORY)
+        .put(TaskConfig.GROUPER_FACTORY(), TEST_TASK_GROUPER_FACTORY)
+        .put(JobCoordinatorConfig.JOB_COORDINATOR_FACTORY, TEST_JOB_COORDINATOR_FACTORY)
+        .put(JobConfig.PROCESSOR_ID(), processorId)
+        .put(ApplicationConfig.APP_NAME, appName)
+        .put(ApplicationConfig.APP_ID, appId)
+        .put(String.format("systems.%s.samza.factory", systemName), TEST_SYSTEM_FACTORY)
+        .put(JobConfig.JOB_NAME(), TEST_JOB_NAME)
+        .build();
+    Map<String, String> applicationConfig = Maps.newHashMap(samzaContainerConfig);
+    applicationConfig.putAll(StandaloneTestUtils.getKafkaSystemConfigs(systemName, bootstrapServers(), zkConnect(), null, StandaloneTestUtils.SerdeAlias.STRING, true));
+    return new ApplicationConfig(new MapConfig(applicationConfig));
+  }
+
+  @Test
+  public void shouldReElectLeaderWhenLeaderDies() throws InterruptedException {
+    // Set up kafka topics.
+    publishKafkaEvents(inputKafkaTopic, NUM_KAFKA_EVENTS, PROCESSOR_IDS[0]);
+
+    // Create stream applications.
+    CountDownLatch processedMessagesLatch1 = new CountDownLatch(1);
+    CountDownLatch processedMessagesLatch2 = new CountDownLatch(1);
+    CountDownLatch processedMessagesLatch3 = new CountDownLatch(1);
+
+    StreamApplication streamApp1 = new TestStreamApplication(inputKafkaTopic, outputKafkaTopic, processedMessagesLatch1, null);
+    StreamApplication streamApp2 = new TestStreamApplication(inputKafkaTopic, outputKafkaTopic, processedMessagesLatch2, null);
+    StreamApplication streamApp3 = new TestStreamApplication(inputKafkaTopic, outputKafkaTopic, processedMessagesLatch3, null);
+
+    // Run stream applications.
+    applicationRunner1.run(streamApp1);
+    applicationRunner2.run(streamApp2);
+    applicationRunner3.run(streamApp3);
+
+    // Wait until all processors have processed a message.
+    processedMessagesLatch1.await();
```

Comment:
Yes, if you're asking i'm defining the rule to terminate tests after all successful assertions.

Can use SystemConsumer with EOS(Injecting custom message list will be hard due to reflection based SystemFactory instantiation). Will help here and other StreamApp tests. Rule can be removed in that case.